soc: arm: stm32g4: power management support

fouge opened the pull request:
Add support for low power modes (`STOP0` and `STOP1`) on STM32G4 series.
Based on STM32G0 implementation.

fouge committed:
soc: arm: stm32g4: PM support

Add power management support for STM32G4 series.
Supported modes include STOP0 and STOP1.

Signed-off-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

fouge commented:
@erwango any comment?

erwango commented on soc/arm/st_stm32/stm32g4/power.c:
```diff
@@ -0,0 +1,98 @@
+	/* keep in mind that debugging draws a lot of power */
+#ifdef CONFIG_DEBUG
+	LL_DBGMCU_EnableDBGStopMode();
+#else
+	LL_DBGMCU_DisableDBGStopMode();
```
Disable is usually not performed in that case, but I ack this is probably a good thing, as it will disable bit systematically at boot and avoid surprises that could be seen when using openocd (which sets this bit during flash) or flashing a debug disabled SW back to back with a debug enabled SW.
We probably need to extend this to other series (after testing though, as this DBGMCU IP block implementation is full of surprises).

fouge replied:
I am not sure what you want me to do here?

erwango replied:
Nothing. This work for me to apply this on other series.

erwango commented on soc/arm/st_stm32/stm32g4/power.c:
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2021 STMicroelectronics.
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
```
This doesn't look accurate.

fouge replied:
fixed
this file is still based on G0 implementation so I kept the previous copyright notices and added a new one

erwango reviewed (changes requested):
This looks globally alright, but please add PM support to one of the in tree G4 boards in order to ensure CI compilation and test of related code (power.c and lptim driver on g4 target).
Ideally `nucleo_g474re` would be great, as it is tested daily on our test bench.

fouge committed:
boards: arm: nucleo_g474re & b_g474e_dpow1 PM support

Add STM32G4x boards support for Power Management with 2 low power
modes: STOP0 and STOP1. LPTIM used as clock source.

Signed-off-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

fouge commented:
@erwango I added PM support for the nucleo_g474re and b_g474e_dpow1
I've been able to test on the `b_g474e_dpow1` only as this is the only dev board I have.

# Tests

## Blinky

From `zephyr/samples/boards/stm32/power_mgmt/blinky`: I can see the LED blink every 2 seconds.

## Tickless concept

Build command:
```
west build -b b_g474e_dpow1
```

Results:
```
2022-12-22 10:50:10.813 ===================================================================
2022-12-22 10:50:10.815 PROJECT EXECUTION SUCCESSFUL
2022-12-22 10:51:45.057 *** Booting Zephyr OS build zephyr-v3.2.0-2633-gc3ffb855396c ***
2022-12-22 10:51:45.060 Running TESTSUITE tickless_concept
2022-12-22 10:51:45.066 ===================================================================
2022-12-22 10:51:45.069 START - test_tickless_slice
2022-12-22 10:51:45.177 elapsed slice 110, expected: <100, 110>
2022-12-22 10:51:45.278 elapsed slice 100, expected: <100, 110>
2022-12-22 10:51:45.378 elapsed slice 100, expected: <100, 110>
2022-12-22 10:51:45.479 elapsed slice 100, expected: <100, 110>
2022-12-22 10:51:45.684  PASS - test_tickless_slice in 0.607 seconds
2022-12-22 10:51:45.691 ===================================================================
2022-12-22 10:51:45.693 START - test_tickless_sysclock
2022-12-22 10:51:45.912 time 640, 850
2022-12-22 10:51:46.125 time 860, 1060
2022-12-22 10:51:46.130  PASS - test_tickless_sysclock in 0.426 seconds
2022-12-22 10:51:46.136 ===================================================================
2022-12-22 10:51:46.139 TESTSUITE tickless_concept succeeded
2022-12-22 10:51:46.139
2022-12-22 10:51:46.142 ------ TESTSUITE SUMMARY START ------
2022-12-22 10:51:46.143
2022-12-22 10:51:46.152 SUITE PASS - 100.00% [tickless_concept]: pass = 2, fail = 0, skip = 0, total = 2 duration = 1.033 seconds
2022-12-22 10:51:46.158  - PASS - [tickless_concept.test_tickless_slice] duration = 0.607 seconds
2022-12-22 10:51:46.165  - PASS - [tickless_concept.test_tickless_sysclock] duration = 0.426 seconds
2022-12-22 10:51:46.165
2022-12-22 10:51:46.168 ------ TESTSUITE SUMMARY END ------
2022-12-22 10:51:46.169
2022-12-22 10:51:46.175 ===================================================================
2022-12-22 10:51:46.177 PROJECT EXECUTION SUCCESSFUL
```

## Timer API

Build command with `prj_tickless.conf`:
```
west build -b b_g474e_dpow1 -- -DCONF_FILE="prj_tickless.conf"
```

Results:
```
2022-12-22 10:57:03.364 *** Booting Zephyr OS build zephyr-v3.2.0-2841-g53c3040a996d ***
2022-12-22 10:57:03.366 Running TESTSUITE timer_api
2022-12-22 10:57:03.372 ===================================================================
2022-12-22 10:57:03.374 START - test_sleep_abs
2022-12-22 10:57:03.391  PASS - test_sleep_abs in 0.014 seconds
2022-12-22 10:57:03.397 ===================================================================
2022-12-22 10:57:03.400 START - test_time_conversions
2022-12-22 10:57:03.455  PASS - test_time_conversions in 0.052 seconds
2022-12-22 10:57:03.461 ===================================================================
2022-12-22 10:57:03.463 START - test_timeout_abs
2022-12-22 10:57:03.468  PASS - test_timeout_abs in 0.002 seconds
2022-12-22 10:57:03.474 ===================================================================
2022-12-22 10:57:03.477 START - test_timer_duration_period
2022-12-22 10:57:03.810  PASS - test_timer_duration_period in 0.326 seconds
2022-12-22 10:57:03.816 ===================================================================
2022-12-22 10:57:03.818 START - test_timer_expirefn_null
2022-12-22 10:57:04.151  PASS - test_timer_expirefn_null in 0.326 seconds
2022-12-22 10:57:04.157 ===================================================================
2022-12-22 10:57:04.159 START - test_timer_k_define
2022-12-22 10:57:04.869  PASS - test_timer_k_define in 0.702 seconds
2022-12-22 10:57:04.875 ===================================================================
2022-12-22 10:57:04.878 START - test_timer_period_0
2022-12-22 10:57:05.085  PASS - test_timer_period_0 in 0.202 seconds
2022-12-22 10:57:05.091 ===================================================================
2022-12-22 10:57:05.094 START - test_timer_period_k_forever
2022-12-22 10:57:05.301  PASS - test_timer_period_k_forever in 0.202 seconds
2022-12-22 10:57:05.307 ===================================================================
2022-12-22 10:57:05.310 START - test_timer_periodicity
2022-12-22 10:57:05.569  PASS - test_timer_periodicity in 0.252 seconds
2022-12-22 10:57:05.575 ===================================================================
2022-12-22 10:57:05.577 START - test_timer_remaining
2022-12-22 10:57:05.632  PASS - test_timer_remaining in 0.051 seconds
2022-12-22 10:57:05.638 ===================================================================
2022-12-22 10:57:05.641 START - test_timer_restart
2022-12-22 10:57:05.922  PASS - test_timer_restart in 0.276 seconds
2022-12-22 10:57:05.928 ===================================================================
2022-12-22 10:57:05.931 START - test_timer_status_get
2022-12-22 10:57:05.935  PASS - test_timer_status_get in 0.001 seconds
2022-12-22 10:57:05.941 ===================================================================
2022-12-22 10:57:05.945 START - test_timer_status_get_anytime
2022-12-22 10:57:06.278  PASS - test_timer_status_get_anytime in 0.326 seconds
2022-12-22 10:57:06.284 ===================================================================
2022-12-22 10:57:06.286 START - test_timer_status_sync
2022-12-22 10:57:06.646  PASS - test_timer_status_sync in 0.352 seconds
2022-12-22 10:57:06.652 ===================================================================
2022-12-22 10:57:06.654 START - test_timer_user_data
2022-12-22 10:57:06.964  PASS - test_timer_user_data in 0.301 seconds
2022-12-22 10:57:06.970 ===================================================================
2022-12-22 10:57:06.972 TESTSUITE timer_api succeeded
2022-12-22 10:57:06.973
2022-12-22 10:57:06.976 ------ TESTSUITE SUMMARY START ------
2022-12-22 10:57:06.976
2022-12-22 10:57:06.985 SUITE PASS - 100.00% [timer_api]: pass = 15, fail = 0, skip = 0, total = 15 duration = 3.385 seconds
2022-12-22 10:57:06.990  - PASS - [timer_api.test_sleep_abs] duration = 0.014 seconds
2022-12-22 10:57:06.996  - PASS - [timer_api.test_time_conversions] duration = 0.052 seconds
2022-12-22 10:57:07.002  - PASS - [timer_api.test_timeout_abs] duration = 0.002 seconds
2022-12-22 10:57:07.008  - PASS - [timer_api.test_timer_duration_period] duration = 0.326 seconds
2022-12-22 10:57:07.015  - PASS - [timer_api.test_timer_expirefn_null] duration = 0.326 seconds
2022-12-22 10:57:07.021  - PASS - [timer_api.test_timer_k_define] duration = 0.702 seconds
2022-12-22 10:57:07.027  - PASS - [timer_api.test_timer_period_0] duration = 0.202 seconds
2022-12-22 10:57:07.033  - PASS - [timer_api.test_timer_period_k_forever] duration = 0.202 seconds
2022-12-22 10:57:07.039  - PASS - [timer_api.test_timer_periodicity] duration = 0.252 seconds
2022-12-22 10:57:07.045  - PASS - [timer_api.test_timer_remaining] duration = 0.051 seconds
2022-12-22 10:57:07.051  - PASS - [timer_api.test_timer_restart] duration = 0.276 seconds
2022-12-22 10:57:07.057  - PASS - [timer_api.test_timer_status_get] duration = 0.001 seconds
2022-12-22 10:57:07.064  - PASS - [timer_api.test_timer_status_get_anytime] duration = 0.326 seconds
2022-12-22 10:57:07.070  - PASS - [timer_api.test_timer_status_sync] duration = 0.352 seconds
2022-12-22 10:57:07.076  - PASS - [timer_api.test_timer_user_data] duration = 0.301 seconds
2022-12-22 10:57:07.076
2022-12-22 10:57:07.079 ------ TESTSUITE SUMMARY END ------
2022-12-22 10:57:07.080
2022-12-22 10:57:07.086 ===================================================================
2022-12-22 10:57:07.088 PROJECT EXECUTION SUCCESSFUL
```

erwango reviewed (approved):
Thanks for the detailed report.

fouge commented:
@erwango 
the serial_wakeup test could be done on Nucleo boards depending on whether the LSE is enabled in your test boards.

<img width="713" alt="image" src="https://user-images.githubusercontent.com/1217771/209114734-93dfd14a-e79e-4579-a02c-432d9606debd.png">

Should I try to propose a working example even though I won't be able to test? 😬

erwango commented:
> @erwango the serial_wakeup test could be done on Nucleo boards depending on whether the LSE is enabled in your test boards.
> 
> <img alt="image" width="713" src="https://user-images.githubusercontent.com/1217771/209114734-93dfd14a-e79e-4579-a02c-432d9606debd.png">
> 
> Should I try to propose a working example even though I won't be able to test? grimacing

That would be nice. We'll do the test (using nucleo_g474re). But please do it in another PR.
